### PR TITLE
Adds support for recompilation of views/pages

### PR DESCRIPTION
**Pull request by rynowak**
See the individual commits for details

### CODE REVIEW

**Comment by rynowak** [comment]
Note, I can't test this end to end yet, and the functional tests are skipped. Waiting on a Razor change to flow through the build.

File: src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ViewsFeatureProvider.cs [hunk 1]
```diff
@@ -19,26 +18,79 @@ public class ViewsFeatureProvider : IApplicationFeatureProvider<ViewsFeature>
...
+        }
 
-                    feature.ViewDescriptors.Add(viewDescriptor);
+        private ICollection<(RazorCompiledItem item, RazorViewAttribute attribute)> Merge(
+            IReadOnlyList<RazorCompiledItem> items, 
+            IEnumerable<RazorViewAttribute> attributes)
+        {
+            // This code is a intentionally defensive. We assume that it's possible to have duplicates
+            // of attributes, and also items that have a single kind of metadata, but not the other.
+            var dictionary = new Dictionary<string, (RazorCompiledItem item, RazorViewAttribute attribute)>();
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ViewsFeatureProvider.cs]: `StringComparer.Ordinal`

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/ChecksumValidator.cs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            for (var i = 0; i < checksums.Count; i++)
+            {
+                var checksum = checksums[i];
+                if (string.Equals(item.Identifier, checksum.Identifier, StringComparison.OrdinalIgnoreCase))
+                {
+                    // Ignore primary checksum on this pass.
+                    continue;
+                }
+
+                projectItem = project.GetItem(checksum.Identifier);
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/ChecksumValidator.cs]: `var importItem`

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/RazorViewCompiler.cs [hunk 3]
```diff
@@ -109,103 +119,193 @@ public Task<CompiledViewDescriptor> CompileAsync(string relativePath)
...
+                SupportsCompilation = true,
+
+                Descriptor = precompiledView, // This might be used, if the checksums match.
+
+                // Used to validate and recompile
+                NormalizedPath = normalizedPath,
+                ExpirationTokens = new List<IChangeToken>(),
+            };
+            
+            var checksums = precompiledView.Item?.GetChecksumMetadata() ?? Array.Empty<IRazorSourceChecksumMetadata>();
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/RazorViewCompiler.cs]: `precompiledView.Item` is never null here.

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/RazorViewCompiler.cs [hunk 4]
```diff
@@ -109,103 +119,193 @@ public Task<CompiledViewDescriptor> CompileAsync(string relativePath)
...
+                    Descriptor = new CompiledViewDescriptor()
+                    {
+                        RelativePath = normalizedPath,
+                        ExpirationTokens = expirationTokens,
+                    },
+
+                    // We can try again if the file gets created.
+                    ExpirationTokens = expirationTokens,
+                };
+
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/RazorViewCompiler.cs]: Extra line

File: test/Microsoft.AspNetCore.Mvc.Razor.Test/Compilation/ViewsFeatureProviderTest.cs [hunk 5]
```diff
@@ -36,7 +37,29 @@ public void PopulateFeature_ReturnsViewsFromAllAvailableApplicationParts()
...
+
+            var items = new Dictionary<AssemblyPart, IReadOnlyList<RazorCompiledItem>>
+            {
+                {
+                    part1,
+                    new[]
+                    {
+                        new TestRazorCompiledItem(typeof(object), "mvc.1.0.view", "/Views/test/Index.cshtml", new object[]{ }),
+
+                        // This one doesn't have a RazorViewAttribute
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Razor.Test/Compilation/ViewsFeatureProviderTest.cs]: I don't follow this comment. Neither does `object` or `string`?

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/ChecksumValidator.cs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+            // The checksum that matches 'Item.Identity' in this list is significant. That represents the main file.
+            //
+            // We don't really care about the validation unless the main file exists. This is because we expect
+            // most sites to have some _ViewImports in common location. That means that in the case you're
+            // using views from a 3rd party library, you'll always have **some** conflicts. 
+            //
+            // The presence of the main file with the same content is a very strong signal that you're in a
+            // development scenario.
+            var primaryChecksum = checksums
+                .Where(c => string.Equals(item.Identifier, c.Identifier, StringComparison.OrdinalIgnoreCase))
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/ChecksumValidator.cs]: `.FirstOrDefault(c =>...)`

File: src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ViewsFeatureProvider.cs (hunk 1 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ViewsFeatureProvider.cs]: OrdinalIgnoreCase

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/ChecksumValidator.cs (hunk 2 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.Razor/Internal/ChecksumValidator.cs]: meh

File: test/Microsoft.AspNetCore.Mvc.Razor.Test/Compilation/ViewsFeatureProviderTest.cs (hunk 5 above)
Comment (rynowak) [reply test/Microsoft.AspNetCore.Mvc.Razor.Test/Compilation/ViewsFeatureProviderTest.cs]: This item doesn't have a RazorViewAttribute.

**Pull request merged**